File: packages/payment/ios/Plugin/PaymentSheet/PaymentSheetExecutor.swift
import Foundation
import Capacitor
import StripePaymentSheet

class PaymentSheetExecutor: NSObject {
    weak var plugin: StripePlugin?
    var paymentSheet: PaymentSheet?
    var continueConfirmHandler: CheckedContinuation<String, Error>? = nil

    func createPaymentSheet(_ call: CAPPluginCall) {
        var configuration = PaymentSheet.Configuration()

        if let merchantDisplayName = call.getString("merchantDisplayName") {
            configuration.merchantDisplayName = merchantDisplayName
        }

        if let returnURL = call.getString("returnURL") {
            configuration.returnURL = returnURL
        }

        if let style = parse(style: call.getString("style")) {
            configuration.style = style
        }

        if let applePayMerchantId = call.getString("applePayMerchantId"), call.getBool("enableApplePay", false) {
            configuration.applePay = .init(
                merchantId: applePayMerchantId,
                merchantCountryCode: call.getString("countryCode", "US")
            )
        }

        let customerId = call.getString("customerId")
        let customerEphemeralKeySecret = call.getString("customerEphemeralKeySecret")

        if let customerId = customerId, let customerEphemeralKeySecret = customerEphemeralKeySecret {
            configuration.customer = .init(id: customerId, ephemeralKeySecret: customerEphemeralKeySecret)
        } else if customerId != nil || customerEphemeralKeySecret != nil {
            let errorText = "Invalid Params. When you set customerId, you must set customerEphemeralKeySecret."
            self.plugin?.notifyListeners(PaymentSheetEvents.FailedToLoad.rawValue, data: ["error": errorText])
            call.reject(errorText)
            return
        }

        let confirmOnServer = call.getBool("confirmOnServer") ?? false
        if !confirmOnServer {
            let paymentIntentClientSecret = call.getString("paymentIntentClientSecret")
            let setupIntentClientSecret = call.getString("setupIntentClientSecret")

            if let setupIntentClientSecret = setupIntentClientSecret {
                self.paymentSheet = PaymentSheet(
                    setupIntentClientSecret: setupIntentClientSecret,
                    configuration: configuration
                )
            } else if let paymentIntentClientSecret = paymentIntentClientSecret {
                self.paymentSheet = PaymentSheet(
                    paymentIntentClientSecret: paymentIntentClientSecret,
                    configuration: configuration
                )
            } else {
                let errorText = "Invalid Params. this method require paymentIntentClientSecret or setupIntentClientSecret."
                self.plugin?.notifyListeners(PaymentSheetEvents.FailedToLoad.rawValue, data: ["error": errorText])
                call.reject(errorText)
                return
            }
        } else {
            let amount = call.getInt("amount") // Required for mode = .payment.
            let currency = call.getString("currency") // Required.
            let setupFutureUsage = parse(setupFutureUse: call.getString("setupFutureUsage")) // Required for mode = .setup, optional for mode = .payment.
            let paymentMethodTypes = call.getArray("paymentMethodTypes", String.self) // Optional.
            let onBehalfOf = call.getString("onBehalfOf") // Optional.

            guard let currency = currency else {
                call.reject("Invalid params. If confirmOnServer is set, currency must also be set.")
                return
            }

            guard let mode = (amount.map { amount in
                let captureMethod = parse(captureMethod: call.getString("captureMethod")) ?? .automatic
                return PaymentSheet.IntentConfiguration.Mode.payment(
                    amount: amount,
                    currency: currency,
                    setupFutureUsage: setupFutureUsage,
                    captureMethod: captureMethod
                )
            } ?? setupFutureUsage.map { setupFutureUsage in
                return PaymentSheet.IntentConfiguration.Mode.setup(
                    currency: currency,
                    setupFutureUsage: setupFutureUsage
                )
            }) else {
                call.reject("Invalid params. Either amount or setupFutureUsage must be set.")
                return
            }

            let intentConfiguration = PaymentSheet.IntentConfiguration(
                mode: mode,
                paymentMethodTypes: paymentMethodTypes,
                onBehalfOf: onBehalfOf,
                confirmHandler: self.confirmHandler(_:_:_:)
            )

            self.paymentSheet = PaymentSheet(
                intentConfiguration: intentConfiguration,
                configuration: configuration
            )
        }

        self.plugin?.notifyListeners(PaymentSheetEvents.Loaded.rawValue, data: [:])
        call.resolve()
    }

    private func parse(style: String?) -> PaymentSheet.UserInterfaceStyle? {
        switch style {
        case "alwaysLight":
            return .alwaysLight
        case "alwaysDark":
            return .alwaysDark
        case "automatic":
            return .automatic
        default:
            return nil
        }
    }

    private func parse(setupFutureUse: String?) -> PaymentSheet.IntentConfiguration.SetupFutureUsage? {
        switch setupFutureUse {
        case "offSession":
            return .offSession
        case "onSession":
            return .onSession
        default:
            return nil
        }
    }

    private func parse(captureMethod: String?) -> PaymentSheet.IntentConfiguration.CaptureMethod? {
        switch captureMethod {
        case "automatic":
            return .automatic
        case "manual":
            return .manual
        case "automaticAsync":
            return .automaticAsync
        default:
            return nil
        }
    }

    func presentPaymentSheet(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            if let rootViewController = self.plugin?.getRootVC() {
                self.paymentSheet?.present(from: rootViewController) { paymentResult in
                    // MARK: Handle the payment result
                    switch paymentResult {
                    case .completed:
                        self.plugin?.notifyListeners(PaymentSheetEvents.Completed.rawValue, data: [:])
                        call.resolve(["paymentResult": PaymentSheetEvents.Completed.rawValue])
                    case .canceled:
                        self.plugin?.notifyListeners(PaymentSheetEvents.Canceled.rawValue, data: [:])
                        call.resolve(["paymentResult": PaymentSheetEvents.Canceled.rawValue])
                    case .failed(let error):
                        self.plugin?.notifyListeners(PaymentSheetEvents.Failed.rawValue, data: ["error": error.localizedDescription])
                        call.resolve(["paymentResult": PaymentSheetEvents.Failed.rawValue])
                    }
                }
            }
        }
    }

    func completeConfirmPaymentSheet(_ call: CAPPluginCall) {
        guard let continueConfirmHandler = continueConfirmHandler else {
            call.reject("Invalid state. Must be called inside the confirmPaymentSheet callback.")
            return
        }

        if let error = call.getString("error") {
            continueConfirmHandler.resume(throwing: StringError(message: error))
            call.resolve()
            return
        }

        guard let clientSecret = call.getString("clientSecret") else {
            continueConfirmHandler.resume(throwing: StringError(message: "Invalid Param. Missing clientSecret or error."))
            call.reject("Invalid Param. Missing clientSecret or error.")
            return
        }

        continueConfirmHandler.resume(returning: clientSecret)
        call.resolve()
    }

    private func confirmHandler(
        _ paymentMethod: STPPaymentMethod,
        _ shouldSavePaymentMethod: Bool,
        _ intentCreationCallback: @escaping ((Result<String, Error>) -> Void)
    ) -> Void {
        Task {
            do {
                let result = try await withCheckedThrowingContinuation { continuation in
                    self.continueConfirmHandler = continuation
                    self.plugin?.notifyListeners(PaymentSheetEvents.ConfirmOnServer.rawValue, data: [
                        "paymentMethod": [
                            "id": paymentMethod.stripeId,
                        ],
                        "shouldSavePaymentMethod": shouldSavePaymentMethod,
                    ])
                }
                intentCreationCallback(.success(result))
            } catch {
                intentCreationCallback(.failure(error))
            }
        }
    }
}

File: packages/payment/android/src/main/java/com/getcapacitor/community/stripe/paymentsheet/PaymentSheetExecutor.kt
package com.getcapacitor.community.stripe.paymentsheet

import android.app.Activity
import android.content.Context
import androidx.core.util.Supplier
import com.getcapacitor.Bridge
import com.getcapacitor.JSObject
import com.getcapacitor.Logger
import com.getcapacitor.PluginCall
import com.getcapacitor.community.stripe.models.Executor
import com.google.android.gms.common.util.BiConsumer
import com.stripe.android.model.PaymentMethod
import com.stripe.android.paymentsheet.CreateIntentResult
import com.stripe.android.paymentsheet.PaymentSheet
import com.stripe.android.paymentsheet.PaymentSheetResult
import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class ErrorException(message: String) : Exception(message)

class PaymentSheetExecutor(
        contextSupplier: Supplier<Context?>,
        activitySupplier: Supplier<Activity?>,
        notifyListenersFunction: BiConsumer<String?, JSObject?>,
        pluginLogTag: String?
) : Executor(contextSupplier, activitySupplier, notifyListenersFunction, pluginLogTag, "PaymentSheetExecutor") {
    var paymentSheet: PaymentSheet? = null
    private val emptyObject = JSObject()
    private var confirmOnServer: Boolean = false
    private var intentConfiguration: PaymentSheet.IntentConfiguration? = null
    private var paymentConfiguration: PaymentSheet.Configuration? = null
    private var paymentIntentClientSecret: String? = null
    private var setupIntentClientSecret: String? = null

    private var continueConfirm: Continuation<String>? = null


    init {
        this.contextSupplier = contextSupplier
    }

    fun createPaymentSheet(call: PluginCall) {
        val customerId = call.getString("customerId")
        val customerEphemeralKeySecret = call.getString("customerEphemeralKeySecret")
        if (customerId != null && customerEphemeralKeySecret == null) {
            val errorText = "Invalid Params. When you set customerId, you must set customerEphemeralKeySecret."
            notifyListenersFunction.accept(PaymentSheetEvents.FailedToLoad.webEventName, JSObject().put("error", errorText))
            call.reject(errorText)
            return
        }

        val merchantDisplayName = call.getString("merchantDisplayName") ?: ""
        val enableGooglePay = call.getBoolean("enableGooglePay") ?: false
        val customer: PaymentSheet.CustomerConfiguration? = customerId?.let { cId ->
            customerEphemeralKeySecret?.let { cSecret ->
                PaymentSheet.CustomerConfiguration(cId, cSecret)
            }
        }
        if (!enableGooglePay) {
            paymentConfiguration = PaymentSheet.Configuration(merchantDisplayName, customer)
        } else {
            val googlePayEnvironment = call.getBoolean("GooglePayIsTesting") ?: false
            val environment = if (googlePayEnvironment)
                PaymentSheet.GooglePayConfiguration.Environment.Test
            else
                PaymentSheet.GooglePayConfiguration.Environment.Production

            paymentConfiguration = PaymentSheet.Configuration(
                    merchantDisplayName,
                    customer,
                    PaymentSheet.GooglePayConfiguration(
                            environment,
                            call.getString("countryCode") ?: "US"
                    )
            )
        }

        confirmOnServer = call.getBoolean("confirmOnServer") ?: false
        if (!confirmOnServer) {
            paymentIntentClientSecret = call.getString("paymentIntentClientSecret")
            setupIntentClientSecret = call.getString("setupIntentClientSecret")
            if (paymentIntentClientSecret == null && setupIntentClientSecret == null) {
                val errorText = "Invalid Params. This method require paymentIntentClientSecret or setupIntentClientSecret."
                notifyListenersFunction.accept(PaymentSheetEvents.FailedToLoad.webEventName, JSObject().put("error", errorText))
                call.reject(errorText)
                return
            }
        } else {
            val amount = call.getLong("amount") ?: call.getInt("amount")?.toLong() // Required for Mode.Payment.
            val currency = call.getString("currency") // Required.
            val setupFutureUsage = parseSetupFutureUse(call.getString("setupFutureUsage")) // Required for Mode.Setup, optional for Mode.Payment.
            val paymentMethodTypes = call.getArray("paymentMethodTypes")?.toList<String>() ?: emptyList() // Optional.
            val onBehalfOf = call.getString("onBehalfOf") // Optional.

            if (currency == null) {
                call.reject("Invalid params. If confirmOnServer is set, currency must also be set.")
                return
            }

            val mode = if (amount != null) {
                val captureMethod = parseCaptureMethod(call.getString("captureMethod")) ?: PaymentSheet.IntentConfiguration.CaptureMethod.Automatic
                PaymentSheet.IntentConfiguration.Mode.Payment(
                        amount,
                        currency,
                        setupFutureUsage,
                        captureMethod
                )
            } else if (setupFutureUsage != null) {
                PaymentSheet.IntentConfiguration.Mode.Setup(
                        currency,
                        setupFutureUsage
                )
            } else {
                call.reject("Invalid params. Either amount or setupFutureUsage must be set.")
                return
            }
            intentConfiguration = PaymentSheet.IntentConfiguration(
                    mode,
                    paymentMethodTypes,
                    onBehalfOf
            )
        }
        notifyListenersFunction.accept(PaymentSheetEvents.Loaded.webEventName, emptyObject)
        call.resolve()
    }

    private fun parseSetupFutureUse(setupFutureUsage: String?): PaymentSheet.IntentConfiguration.SetupFutureUse? {
        return when (setupFutureUsage) {
            "offSession" -> PaymentSheet.IntentConfiguration.SetupFutureUse.OffSession
            "onSession" -> PaymentSheet.IntentConfiguration.SetupFutureUse.OnSession
            else -> null
        }
    }

    private fun parseCaptureMethod(captureMethod: String?): PaymentSheet.IntentConfiguration.CaptureMethod? {
        return when (captureMethod) {
            "automatic" -> PaymentSheet.IntentConfiguration.CaptureMethod.Automatic
            "manual" -> PaymentSheet.IntentConfiguration.CaptureMethod.Manual
            "automaticAsync" -> PaymentSheet.IntentConfiguration.CaptureMethod.AutomaticAsync
            else -> null
        }
    }

    fun presentPaymentSheet(call: PluginCall) {
        try {
            if (intentConfiguration != null) {
                paymentSheet!!.presentWithIntentConfiguration(intentConfiguration!!, paymentConfiguration)
            } else {
                if (paymentIntentClientSecret != null) {
                    paymentSheet!!.presentWithPaymentIntent(paymentIntentClientSecret!!, paymentConfiguration)
                } else {
                    paymentSheet!!.presentWithSetupIntent(setupIntentClientSecret!!, paymentConfiguration)
                }
            }
        } catch (ex: Exception) {
            call.reject(ex.localizedMessage, ex)
        }
    }

    fun completeConfirmPaymentSheet(call: PluginCall) {
        try {
            val continueConfirm = continueConfirm
            if (continueConfirm == null) {
                call.reject("Invalid state. Must be called inside the finalizePaymentSheet callback.")
                return
            }

            val error = call.getString("error")
            if (error != null) {
                continueConfirm.resumeWithException(ErrorException(error))
                call.resolve()
                return
            }

            val clientSecret = call.getString("clientSecret")
            if (clientSecret == null) {
                continueConfirm.resumeWithException(ErrorException("Invalid Param. Missing clientSecret or error."))
                call.reject("Invalid Param. Missing clientSecret or error.")
            } else {
                continueConfirm.resume(clientSecret)
            }

            call.resolve()
        } catch (ex: Exception) {
            call.reject(ex.localizedMessage, ex)
        }
    }

    suspend fun onConfirm(bridge: Bridge, callbackId: String?, paymentMethod: PaymentMethod, shouldSavePaymentMethod: Boolean): CreateIntentResult {
        Logger.info("ConfirmIntentCallback called")
        return try {
            val clientSecret = suspendCoroutine { cont ->
                continueConfirm = cont
                notifyListenersFunction.accept(PaymentSheetEvents.ConfirmOnServer.webEventName, JSObject()
                        .put("paymentMethod", JSObject().put("id", paymentMethod.id))
                        .put("shouldSavePaymentMethod", shouldSavePaymentMethod)
                )
            }
            CreateIntentResult.Success(clientSecret)
        } catch (error: ErrorException) {
            CreateIntentResult.Failure(error, error.message)
        } catch (ex: Exception) {
            CreateIntentResult.Failure(ex, ex.localizedMessage)
        }    }

    fun onPaymentSheetResult(bridge: Bridge, callbackId: String?, paymentSheetResult: PaymentSheetResult) {
        val call = bridge.getSavedCall(callbackId)
        when (paymentSheetResult) {
            is PaymentSheetResult.Canceled -> {
                notifyListenersFunction.accept(PaymentSheetEvents.Canceled.webEventName, emptyObject)
                call.resolve(JSObject().put("paymentResult", PaymentSheetEvents.Canceled.webEventName))
            }

            is PaymentSheetResult.Failed -> {
                notifyListenersFunction.accept(
                        PaymentSheetEvents.Failed.webEventName,
                        JSObject().put("error", paymentSheetResult.error.localizedMessage)
                )
                notifyListenersFunction.accept(PaymentSheetEvents.Failed.webEventName, emptyObject)
                call.resolve(JSObject().put("paymentResult", PaymentSheetEvents.Failed.webEventName))
            }

            is PaymentSheetResult.Completed -> {
                notifyListenersFunction.accept(PaymentSheetEvents.Completed.webEventName, emptyObject)
                call.resolve(JSObject().put("paymentResult", PaymentSheetEvents.Completed.webEventName))
            }
        }
    }
}